### PR TITLE
add exercise 1 solution

### PR DESCRIPTION
**Pull request by stevelee2497**

### CODE REVIEW

File: exercise1/Program.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var dataCenter = "us17";
+            var apiKey = "ed37a130dbadee9671f725ebdf03583a-us17";
+            var listId = "8f5f9903d4";
+            var campaignId = "39494cd8ea";
+
+            Console.WriteLine(GetLists(dataCenter, apiKey, ""));
```
Comment (voquanghoa) [inline exercise1/Program.cs]: Program should enter `email content`, `email to` from user and send the email

File: exercise1/Program.cs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        private static string GetLists(string dataCenter, string apiKey, string listId = "")
+        {
+            var uri = string.Format("https://{0}.api.mailchimp.com/3.0/lists/{1}", dataCenter, listId);
+            try
+            {
+                using (var webClient = new WebClient())
+                {
+                    webClient.Headers.Add("Accept", "application/json");
```
Comment (voquanghoa) [inline exercise1/Program.cs]: Use [Retrofit.net](https://github.com/jordan-thoms/Retrofit.Net)

File: exercise1/IHelper.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using RestSharp;
+using Retrofit.Net.Attributes.Methods;
+using Retrofit.Net.Attributes.Parameters;
+
+namespace exercise1
+{
+    public interface IHelper
```
Comment (voquanghoa) [inline exercise1/IHelper.cs]: Unmeaningful class name.

File: exercise1/IHelper.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        RestResponse<bool> AddEmail([Path("idList")] string id, [Body] object obj);
+
+        [Post("campaigns")]
+        RestResponse<bool> CreateCampaigns([Body] Object obj);
+
+        [Put("campaigns/{idCompaign}/content")]
+        RestResponse<bool> SetContent([Path("idCompaign")] string campaignId, [Body] object obj);
+
+        [Post("campaigns/{idCompaign}/actions/send")]
+        RestResponse<bool> SendCampaign([Path("idCompaign")] string campaignId);
```
Comment (voquanghoa) [inline exercise1/IHelper.cs]: You should split this class into two different classes: `IListService`, `ICampaignService`.

File: exercise1/Program.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            Console.WriteLine(service.CreateCampaigns(campaign).ResponseStatus);
+            // set content for a campaign
+            var content = GetContentObject("Hello there!");
+            Console.WriteLine(service.SetContent(campaignId,content));
+            // send campaign
+            Console.WriteLine(service.SendCampaign(campaignId));
+            Console.ReadKey();
+        }
+        
+        private static IHelper GetService(string url, string apiKey)
```
Comment (voquanghoa) [inline exercise1/Program.cs]: When you split the class `IHelper`, this method should be generic.

File: exercise1/Program.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        {
+            var restClient = new RestClient(url);
+            restClient.Authenticator = new HttpBasicAuthenticator("username", apiKey);
+            var adapter = new RestAdapter(restClient);
+            return adapter.Create<IHelper>();
+        }
+
+        private static Object GetListObject()
```
Comment (voquanghoa) [inline exercise1/Program.cs]: Move this method (and all methods below) to a new class.

File: exercise1/Program.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+using RestSharp;
+using Retrofit.Net;
+
+namespace exercise1
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var dataCenter = "us17";
```
Comment (voquanghoa) [inline exercise1/Program.cs:11]: The program should able to enter an email, an email content from user and send the email.

**Review by voquanghoa — CHANGES_REQUESTED** [review]

File: exercise1/IHelper.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        RestResponse<bool> CreateList([Body] Object obj);
+
+        [Post("lists/{idList}/members")]
+        RestResponse<bool> AddEmail([Path("idList")] string id, [Body] object obj);
+
+        [Post("campaigns")]
+        RestResponse<bool> CreateCampaigns([Body] Object obj);
+
+        [Put("campaigns/{idCompaign}/content")]
+        RestResponse<bool> SetContent([Path("idCompaign")] string campaignId, [Body] object obj);
```
Comment (voquanghoa) [inline exercise1/IHelper.cs]: Should change
- Make the variable name and the place-holder are same together. For example `campaignId` and `idCompaign`
- Specific type for `[Body]` and param name `obj` should be meaningful

File: exercise1/Program.cs (hunk 1 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/Program.cs (hunk 2 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/IHelper.cs (hunk 4 above)
Comment (stevelee2497) [reply exercise1/IHelper.cs]: done

File: exercise1/Program.cs (hunk 7 above)
Comment (stevelee2497) [reply exercise1/Program.cs:11]: done

File: exercise1/Program.cs (hunk 5 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/Program.cs (hunk 6 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/IHelper.cs (hunk 8 above)
Comment (stevelee2497) [reply exercise1/IHelper.cs]: done

File: exercise1/ObjectHelper.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+
+namespace exercise1
+{
+    public class ObjectHelper
+    {
+        public static Object GetListObject()
```
Comment (voquanghoa) [inline exercise1/ObjectHelper.cs]: `Create`, not `Get`

File: exercise1/ObjectHelper.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+
+namespace exercise1
+{
+    public class ObjectHelper
```
Comment (voquanghoa) [inline exercise1/ObjectHelper.cs]: Bad class's name.

File: exercise1/Program.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            var apiKey = "ed37a130dbadee9671f725ebdf03583a-us17";
+            var url = $"https://{dataCenter}.api.mailchimp.com/3.0";
+            // init service
+            var listService = GetService<IListService>(url, apiKey);
+            // create a new list
+            var list = ObjectHelper.GetListObject();
+            Console.WriteLine(listService.CreateList(list).ResponseStatus);
+            // add an email to list
+            var listId = "ea602c2f5c";
```
Comment (voquanghoa) [inline exercise1/Program.cs]: The listId should have the value from the response above

File: exercise1/Program.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            var emailSubcriber = ObjectHelper.GetEmailObject("quoctran2124@gmail.com");
+            Console.WriteLine(listService.AddEmail(listId,emailSubcriber));
+            // create a new campaign
+            var campaignService = GetService<ICampaignService>(url, apiKey);
+            var campaign = ObjectHelper.GetCampaignObject(listId);
+            Console.WriteLine(campaignService.CreateCampaigns(campaign).ResponseStatus);
+            // set content for a campaign
+            var campaignId = "39494cd8ea";
```
Comment (voquanghoa) [inline exercise1/Program.cs]: `campaignId ` should have the value from the request above

File: exercise1/Program.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+            var content = ObjectHelper.GetContentObject("Hello there!");
+            Console.WriteLine(campaignService.SetContent(campaignId,content));
+            // send campaign
+            Console.WriteLine(campaignService.SendCampaign(campaignId));
+            Console.ReadKey();
+        }
+        
+        private static T GetService<T>(string url, string apiKey) where T : class
+        {
+            var restClient = new RestClient(url);
```
Comment (voquanghoa) [inline exercise1/Program.cs:40]: Use `Object initialization`

File: exercise1/ObjectHelper.cs (hunk 9 above)
Comment (stevelee2497) [reply exercise1/ObjectHelper.cs]: done

File: exercise1/ObjectHelper.cs (hunk 10 above)
Comment (stevelee2497) [reply exercise1/ObjectHelper.cs]: done

File: exercise1/Program.cs (hunk 11 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/Program.cs (hunk 12 above)
Comment (stevelee2497) [reply exercise1/Program.cs]: done

File: exercise1/Program.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+            var campaignService = GetService<ICampaignService>(url, apiKey);
+            var campaign = MailChimpObjects.CreateCampaignObject(listId);
+            var campaignResponse = campaignService.CreateCampaigns(campaign);
+            Console.WriteLine($"Create campaign : {campaignResponse.ResponseStatus}");
+            // set content for a campaign
+            var campaignId = campaignResponse.Data.Id;
+            var content = MailChimpObjects.CreateContentObject("Hello there!");
+            Console.WriteLine($"Set content for campaign : {campaignService.SetContent(campaignId, content).ResponseStatus}");
+            // send campaign
+            Console.WriteLine($"Send campaign : {campaignService.SendCampaign(campaignId).ErrorException}");
```
Comment (voquanghoa) [inline exercise1/Program.cs:34]: Doesn't work!!!

**Review by voquanghoa — CHANGES_REQUESTED** [review]

**Pull request merged**